### ISSUE TITLE
molfiles not accounted for in count line of RXN file end up as agents

### ISSUE THREAD

**Issue by uli-f** [cdk/cdk#1132]
The `MDLRXNV2000Reader` adds all molfiles that are not accounted for on the count line of the rxnfile as agents.

Here is a table that shows the behavior of the current implementation:

| No|Reader Mode | count line #reactants | count line #products | count line #agents | #molfiles in rxnfile | #reactants IReaction | #products IReaction | #agents IReaction |
|---|---|---|---|---|---|---|---|---|
| 1|`STRICT` | 1 | 1 | 0 | 4 | 1 | 1 | 2 |
| 2|`STRICT` | 1 | 2 | 0 | 4 | 1 | 2 | 1 |
| 3|`STRICT` | 1 | 2 | 1 | 4 | `CDKException` | `CDKException` | `CDKException` |
| 4|`RELAXED` | 1 | 1 | 0 | 4 | 1 | 1 | 2 |
| 5|`RELAXED` | 1 | 1 | 2 | 4 | 1 | 1 | 2 |
| 6|`RELAXED` | 2 | 1 | 0 | 4 | 2 | 1 | 1 |
| 7|`RELAXED` | 2 | 1 | 1 | 4 | 2 | 1 | 1 |

The numbers above can be easily verified by playing around with the following method:
```
void countline_test() throws IOException, CDKException {
String rxnFile = "$RXN\n" +
        "\n" +
        "      Mrv2403  121320241522\n" +
        "\n" +
        "  2  2\n" +
        "$MOL\n" +
        "\n" +
        "  Mrv2403 12132415222D          \n" +
        "\n" +
        "  1  0  0  0  0  0            999 V2000\n" +
        "   -8.1250    3.3542    0.0000 F   0  0  0  0  0  0  0  0  0  0  0  0\n" +
        "M  END\n" +
        "$MOL\n" +
        "\n" +
        "  Mrv2403 12132415222D          \n" +
        "\n" +
        "  1  0  0  0  0  0            999 V2000\n" +
        "   -5.3958    3.2917    0.0000 Cl  0  0  0  0  0  0  0  0  0  0  0  0\n" +
        "M  END\n" +
        "$MOL\n" +
        "\n" +
        "  Mrv2403 12132415222D          \n" +
        "\n" +
        "  1  0  0  0  0  0            999 V2000\n" +
        "    0.1042    3.3125    0.0000 Br  0  0  0  0  0  0  0  0  0  0  0  0\n" +
        "M  END\n" +
        "$MOL\n" +
        "\n" +
        "  Mrv2403 12132415222D          \n" +
        "\n" +
        "  1  0  0  0  0  0            999 V2000\n" +
        "    2.1667    3.3958    0.0000 I   0  0  0  0  0  0  0  0  0  0  0  0\n" +
        "M  END\n";

IChemObjectBuilder builder = SilentChemObjectBuilder.getInstance();
try (MDLRXNV2000Reader reader = new MDLRXNV2000Reader(new StringReader(rxnFile))) {
        Mode mode = Mode.STRICT;
        reader.setReaderMode(mode);
        IReaction reaction = reader.read(builder.newReaction());
        System.out.println("reader mode=" + mode);
        System.out.println("actual number of reactants=" + reaction.getReactantCount());
        System.out.println("actual number of products=" + reaction.getProductCount());
        System.out.println("actual number of agents=" + reaction.getAgents().getAtomContainerCount());
}
```

Personally, I would expect an exception (or at the very least a logged message) in `RELAXED` mode if the sum of reaction components in the count line and the actual number of molfiles are different from each other.

I don't know enough about what's best in terms of compatibility when it comes to the `STRICT` mode and any additional reaction components present as molfiles, but assume that the current behavior is intended.

@johnmay Does the current implementation (and thus the table above) reflect the expected behavior of the `MDLRXNV2000Reader`?

**Comment by johnmay** [comment]
Note that officially RXN files don't have any agents :-)

**Comment by johnmay** [comment]
So yes strict mode should throw and exception if you get more than expected. Note is correct to throw if strict and agent count is specified since this is not a valid file (agent count is not a thing).

**Comment by uli-f** [comment]
> So yes strict mode should throw and exception if you get more than expected.

sgtm

> Note is correct to throw if strict and agent count is specified since this is not a valid file (agent count is an extension).

Acknowledged.

What about relaxed mode and having more molfiles in the rxnfile than expected given the numbers of the count line:
(1) Exception
(2) log message and add them as agents or
(3) log message and don't add them as agents?

**Comment by johnmay** [comment]
I think don’t log and add, I believe it is that way because the omission of agents this was a common way to do it

**Comment by uli-f** [comment]
changes implemented in #1138
